### PR TITLE
Coda new actions

### PR DESCRIPTION
**Pull request by andrewjschuang**
Actions:

- [ ] Create a Doc | Creates a new Coda doc
- [ ] List Docs | Returns a list of Coda docs accessible by the user
- [ ] Copy Doc | Creates a copy of the specified Coda doc
- [ ] Create a Row | Creates a row in a selected table
- [ ] Update a Row | Updates a specified row in the table
- [ ] Upsert a Row | Creates a new row or updates existing rows if any upsert key columns are provided. When upserting, if multiple rows match the specified key column(s), they will all be updated with the specified value
- [ ] Find Row | Searches for a Coda row in the selected table using a column match search

### CODE REVIEW

**Comment by dylburger** [comment]
- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

File: components/coda/actions/copy-doc/copy-doc.mjs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import coda from "../../coda.app.mjs";
+
+export default {
+  key: "coda_copy-doc",
+  name: "Copy Doc",
+  description: "Creates a copy of the specified Coda doc",
+  version: "0.0.17",
```
Comment (dylburger) [inline components/coda/actions/copy-doc/copy-doc.mjs:7]: This is hard when you're developing, but before you merge, you'll want to reset these to `0.0.1`: https://pipedream.com/docs/components/guidelines/#versioning .

We're planning to introduce a `pd dev` command for actions so that you don't have to constantly version to test.

File: components/coda/actions/copy-doc/copy-doc.mjs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  type: "action",
+  props: {
+    coda,
+    sourceDoc: {
+      propDefinition: [
+        coda,
+        "sourceDoc",
+      ],
+      description: "The doc ID from which to create a copy.",
```
Comment (dylburger) [inline components/coda/actions/copy-doc/copy-doc.mjs:16]: You can remove the "." from the end of this, since it's not a complete sentence

File: components/coda/coda.app.mjs [hunk 3]
```diff
@@ -1,11 +1,158 @@
...
-  propDefinitions: {},
+  propDefinitions: {
+    title: {
+      type: "string",
+      label: "Doc Title",
+      description: "Title of the doc",
+      optional: true,
+    },
+    folderId: {
+      type: "string",
```
Comment (dylburger) [inline components/coda/coda.app.mjs:14]: Is there an API endpoint for listing folders so that we can present this via async options, as well?

File: components/coda/actions/copy-doc/copy-doc.mjs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      optional: false,
```
Comment (dylburger) [inline components/coda/actions/copy-doc/copy-doc.mjs:17]: I'd recommend removing `optional: true` from the original prop definition. Then you don't have to override it here. There are no clear rules re: optionality but I generally don't like to set the `optional` flag in the original prop definition in the app file, and only override with `optional: true` in specific components. 

The main exception is when the prop is _always_ optional across all components. Then you can set `optional: true` in the prop definition.

File: components/coda/actions/copy-doc/copy-doc.mjs (hunk 2 above)
Comment (dylburger) [reply components/coda/actions/copy-doc/copy-doc.mjs:16]: This description is also virtually the same as the one in the prop definition. Maybe remove it here?

File: components/coda/actions/copy-doc/copy-doc.mjs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        "title",
+      ],
+      description: "Title of the newly copied doc. Defaults to 'Untitled'.",
+    },
+    folderId: {
+      propDefinition: [
+        coda,
+        "folderId",
+      ],
+      description: "The ID of the folder within which to copy this doc. Defaults to your \"My docs\" folder in the oldest workspace you joined; this is subject to change. You can get this ID by opening the folder in the docs list on your computer and grabbing the folderId query parameter.",
```
Comment (dylburger) [inline components/coda/actions/copy-doc/copy-doc.mjs:31]: Any references to other props like `folderId` within descriptions should be surrounded with backticks to render them as code.

File: components/coda/actions/list-docs/list-docs.mjs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    },
+    paginate: {
+      propDefinition: [
+        coda,
+        "paginate",
+      ],
+    },
+  },
+  async run() {
+    var params = {
```
Comment (dylburger) [inline components/coda/actions/list-docs/list-docs.mjs:82]: We prefer `const` / `let` over `var`.

File: components/coda/coda.app.mjs [hunk 7]
```diff
@@ -1,11 +1,158 @@
+import axios from "axios";
```
Comment (dylburger) [inline components/coda/coda.app.mjs:1]: Prefer `@pipedream/platform` axios — see the docs from [this guideline](https://pipedream.com/docs/components/guidelines/#use-pipedream-platform-axios-for-all-http-requests)

File: components/coda/actions/list-docs/list-docs.mjs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      query: this.query,
+      sourceDoc: this.sourceDoc,
+      isStarred: this.isStarred,
+      inGallery: this.inGallery,
+      workspaceId: this.workspaceId,
+      folderId: this.folderId,
+      limit: this.limit,
+      pageToken: this.pageToken,
+    };
+    Object.keys(params).forEach((key) => (params[key] === null
```
Comment (dylburger) [inline components/coda/actions/list-docs/list-docs.mjs:94]: Is there a reason you need this? `axios` should strip `null` and `undefined` properties

File: components/coda/actions/list-docs/list-docs.mjs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      || params[key] === undefined
+      || params[key] === "")
+      && delete params[key]);
+
+    var result = await this.coda.listDocs(params);
+
+    if (!this.paginate) {
```
Comment (dylburger) [inline components/coda/actions/list-docs/list-docs.mjs:101]: For now you can remove the `paginate` prop and always paginate / return all items to the user. This is the common use case, and we're trying to remove the complexity of needing to paginate. We may revisit this in the future and expose the pagination / offset params directly to the user.

File: components/coda/actions/list-docs/list-docs.mjs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+      folderId: this.folderId,
+      limit: this.limit,
+      pageToken: this.pageToken,
+    };
+    Object.keys(params).forEach((key) => (params[key] === null
+      || params[key] === undefined
+      || params[key] === "")
+      && delete params[key]);
+
+    var result = await this.coda.listDocs(params);
```
Comment (dylburger) [inline components/coda/actions/list-docs/list-docs.mjs:99]: When you remove the `paginate` prop (see comment below), you can remove this initial call and perhaps convert the `while` loop to a `do...while` loop.

File: components/coda/actions/list-docs/list-docs.mjs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+      return {
+        items: result.items,
+      };
+    }
+
+    var docList = result.items;
+    while (result.nextPageToken) {
+      params["pageToken"] = result.nextPageToken;
+      result = await this.coda.listDocs(params);
+      docList = [
```
Comment (dylburger) [inline components/coda/actions/list-docs/list-docs.mjs:111]: I would recommend simplifying to:

```javascript
docList.push(...result.items)
```

File: components/coda/actions/list-docs/list-docs.mjs [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        ...docList,
+        ...result.items,
+      ];
+    }
+
+    return {
```
Comment (dylburger) [inline components/coda/actions/list-docs/list-docs.mjs:117]: If you rename `docList` to `items` above you can simply `return { items }` here

File: components/coda/coda.app.mjs [hunk 13]
```diff
@@ -1,11 +1,158 @@
...
+      async options () {
+        return this._getKeyValuePair(
+          (await this.listDocs()).items
+        );
+      },
+    },
+    isOwner: {
+      type: "boolean",
+      label: "Is Owner Docs",
+      description: "Show only docs owned by the user.",
```
Comment (dylburger) [inline components/coda/coda.app.mjs:33]: You can remove most of the periods at the end of these prop descriptions

File: components/coda/coda.app.mjs [hunk 14]
```diff
@@ -1,11 +1,158 @@
...
+    paginate: {
+      type: "boolean",
+      label: "Auto-Paginate",
+      description: "By default, list all docs matching search results across all result pages. Set to `false` to limit results to the first page.",
+      optional: true,
+      default: true,
+    },
+  },
   methods: {
+    _getKeyValuePair(list) {
```
Comment (dylburger) [inline components/coda/coda.app.mjs:92]: I would recommend a method name like `_makeOptionsResponse` or something that clarifies that this method should be used to specifically return the array of objects specific to the `options` method.

File: components/coda/coda.app.mjs [hunk 15]
```diff
@@ -1,11 +1,158 @@
...
+      return list.map(
+        (e) => ({
+          label: e.name,
+          value: e.id,
+        })
+      );
+    },
     // this.$auth contains connected account data
     authKeys() {
```
Comment (dylburger) [inline components/coda/coda.app.mjs:101]: If you see these in app files, you can actually remove them. They're included in the app files we programmatically generate for new apps, but once we fill out the app file's code, we can remove the example.

File: components/coda/coda.app.mjs [hunk 16]
```diff
@@ -1,11 +1,158 @@
...
       console.log(Object.keys(this.$auth));
     },
+    /**
+     * Creates a new Coda doc
```
Comment (dylburger) [inline components/coda/coda.app.mjs:105]: Nice JSDocs

File: components/coda/coda.app.mjs [hunk 17]
```diff
@@ -1,11 +1,158 @@
...
+     *
+     * @param {string} title - Title of the new doc
+     * @param {string} folderId - The ID of the folder within to create this
+     * doc
+     * @param {string} [sourceDoc] - An optional doc ID from which to create a
+     * copy
+     * @returns {string} ID of the newly created doc
+     */
+    async createDoc(title, folderId, sourceDoc = "") {
```
Comment (dylburger) [inline components/coda/coda.app.mjs:114]: Does Coda require the empty string be passed?

File: components/coda/coda.app.mjs [hunk 18]
```diff
@@ -1,11 +1,158 @@
...
+      const config = {
+        method: "post",
```
Comment (dylburger) [inline components/coda/coda.app.mjs:116]: Abstract the HTTP request details into a `_makeRequest` method, like here: https://github.com/PipedreamHQ/pipedream/blob/cde8b56751352236da49eafb3bd886c46019f3fb/components/github/github.app.js#L197

That way you're not repeating the base URL, `Authorization` header, etc.

File: components/coda/actions/copy-doc/copy-doc.mjs [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+    },
+  },
+  async run() {
+    return await this.coda.createDoc(
+      this.title,
+      this.folderId,
+      this.sourceDoc,
+    );
```
Comment (dylburger) [inline components/coda/actions/copy-doc/copy-doc.mjs:39]: On action success, we like to include a `$summary` export to tell the user exactly what happened within the action,  for example:

```javascript
async run({ $ }) {
  const resp = await this.coda.createDoc(
    this.title,
    this.folderId,
    this.sourceDoc,
  );
  $.export("$summary", `Copied doc to new doc ${this.title}`)
  return resp
}
```

File: components/coda/actions/copy-doc/copy-doc.mjs [hunk 20]
```diff
@@ -0,0 +1,41 @@
+import coda from "../../coda.app.mjs";
+
+export default {
+  key: "coda_copy-doc",
```
Comment (dylburger) [inline components/coda/actions/copy-doc/copy-doc.mjs:4]: Use `coda-` instead of `coda_`, so this should be `coda-copy-doc`.

We use underscores in place of spaces when the name of the app itself includes them, like in `google_drive-copy-doc`

File: components/coda/actions/copy-doc/copy-doc.mjs [hunk 21]
```diff
@@ -0,0 +1,41 @@
+import coda from "../../coda.app.mjs";
```
Comment (dylburger) [inline components/coda/actions/copy-doc/copy-doc.mjs:1]: Did you run across any errors when developing the actions? e.g. did you observe any 400 errors from Coda on bad inputs? I like to test out edge cases like that and, if the error Coda returns isn't clear, I sometimes like to return my own. See https://pipedream.com/docs/components/guidelines/#error-handling-and-input-validation for more examples.

It's not relevant for all components, but I wanted to mention it if you think the Coda errors can be improved or if there's better input validation we need to do for users.

**Review by dylburger — DISMISSED** [review]
Overall looks good, I added a few comments, mostly re: our guidelines.

File: components/coda/actions/copy-doc/copy-doc.mjs (hunk 1 above)
Comment (andrewjschuang) [reply components/coda/actions/copy-doc/copy-doc.mjs]: Issue addressed

File: components/coda/actions/copy-doc/copy-doc.mjs (hunk 2 above)
Comment (andrewjschuang) [reply components/coda/actions/copy-doc/copy-doc.mjs]: Issue addressed

File: components/coda/actions/copy-doc/copy-doc.mjs (hunk 4 above)
Comment (andrewjschuang) [reply components/coda/actions/copy-doc/copy-doc.mjs]: Issue addressed

File: components/coda/actions/copy-doc/copy-doc.mjs (hunk 5 above)
Comment (andrewjschuang) [reply components/coda/actions/copy-doc/copy-doc.mjs]: Issue addressed

File: components/coda/actions/list-docs/list-docs.mjs (hunk 6 above)
Comment (andrewjschuang) [reply components/coda/actions/list-docs/list-docs.mjs]: Issue addressed

File: components/coda/coda.app.mjs (hunk 7 above)
Comment (andrewjschuang) [reply components/coda/coda.app.mjs]: Issue addressed

File: components/coda/actions/list-docs/list-docs.mjs (hunk 8 above)
Comment (andrewjschuang) [reply components/coda/actions/list-docs/list-docs.mjs]: Yes, there is no need for that. Removed this part.
Issue addressed

File: components/coda/actions/list-docs/list-docs.mjs (hunk 10 above)
Comment (andrewjschuang) [reply components/coda/actions/list-docs/list-docs.mjs]: Issue addressed

File: components/coda/actions/list-docs/list-docs.mjs (hunk 11 above)
Comment (andrewjschuang) [reply components/coda/actions/list-docs/list-docs.mjs]: Issue addressed

File: components/coda/coda.app.mjs (hunk 14 above)
Comment (andrewjschuang) [reply components/coda/coda.app.mjs]: Issue addressed

File: components/coda/coda.app.mjs (hunk 15 above)
Comment (andrewjschuang) [reply components/coda/coda.app.mjs]: Issue addressed

File: components/coda/coda.app.mjs (hunk 17 above)
Comment (andrewjschuang) [reply components/coda/coda.app.mjs]: In the Coda API, the endpoint createDoc is used for creating new docs (with no sourceDocId - I've changed the variable name) and copying docs (with sourceDocId). Not sure if it is better to separate into two methods?

File: components/coda/coda.app.mjs (hunk 18 above)
Comment (andrewjschuang) [reply components/coda/coda.app.mjs:163]: Issue addressed

File: components/coda/actions/copy-doc/copy-doc.mjs (hunk 20 above)
Comment (andrewjschuang) [reply components/coda/actions/copy-doc/copy-doc.mjs]: Issue addressed

File: components/coda/coda.app.mjs (hunk 3 above)
Comment (andrewjschuang) [reply components/coda/coda.app.mjs:24]: Coda API doesn't have an endpoint for listing folders, as seen in the folderId description in the following link.

> The ID of the folder within which to create this doc. Defaults to your "My docs" folder in the oldest workspace you joined; this is subject to change. You can get this ID by opening the folder in the docs list on your computer and grabbing the folderId query parameter.

https://coda.io/developers/apis/v1#operation/createDoc

File: components/coda/actions/copy-doc/copy-doc.mjs (hunk 19 above)
Comment (andrewjschuang) [reply components/coda/actions/copy-doc/copy-doc.mjs]: Issue addressed

File: components/coda/actions/copy-doc/copy-doc.mjs (hunk 21 above)
Comment (andrewjschuang) [reply components/coda/actions/copy-doc/copy-doc.mjs:1]: Issue addressed

File: components/coda/coda.app.mjs (hunk 3 above)
Comment (andrewjschuang) [reply components/coda/coda.app.mjs:25]: This is the same for workspaceIds

File: components/coda/actions/copy-doc/copy-doc.mjs (hunk 1 above)
Comment (dylburger) [reply components/coda/actions/copy-doc/copy-doc.mjs]: Yes that generally should work fine

File: components/coda/actions/list-docs/list-docs.mjs (hunk 9 above)
Comment (andrewjschuang) [reply components/coda/actions/list-docs/list-docs.mjs]: Issue addressed

File: components/coda/coda.app.mjs (hunk 3 above)
Comment (dylburger) [reply components/coda/coda.app.mjs:25]: Ack, thanks

File: components/coda/actions/list-docs/list-docs.mjs [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+      label: "In Gallery Docs",
+      description: "Show only docs visible within the gallery",
+      optional: true,
+    },
+    limit: {
+      propDefinition: [
+        coda,
+        "limit",
+      ],
+    },
```
Comment (js07) [inline components/coda/actions/list-docs/list-docs.mjs:68]: To differentiate the `limit` prop as it's used in this action (to limit the number of docs retrieved over multiple requests) from other actions, I'd suggest creating (and renaming to) a separate `max`/`maxItems` prop (e.g., [`twitter.app.js`](https://github.com/PipedreamHQ/pipedream/blob/9b9500f95e665a9599325b522b51a95334fd91c2/components/twitch/twitch.app.js#L26)).

Also, the `limit` prop's [`max`](https://github.com/PipedreamHQ/pipedream/blob/023b3cbafcabb126d149374449b8e62bbea660ba/components/coda/coda.app.mjs#L131) value, as defined in `coda.app.mjs`, is `50`. To support returning all items to the user, remove the `max` property from the `maxItems` prop, and make it optional.

**Review by js07 — COMMENTED** [review]
Looks good overall! I made some comments related to the Pipedream guidelines and improving usability of some of the actions & props.

File: components/coda/actions/create-doc/create-doc.mjs [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+    },
+  },
+  async run({ $ }) {
+    let data = {
+      title: this.title,
+      folderId: this.folderId,
+    };
+
+    let response = await this.coda.createDoc(data);
+    $.export("$summary", `Created "${response.name}" doc in folderId: "${response.folderId}" and workspaceId: "${response.workspaceId}"`);
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/create-doc/create-doc.mjs:32]: Issue addressed

File: components/coda/actions/create-rows/create-rows.mjs [hunk 24]
```diff
@@ -0,0 +1,70 @@
+import coda from "../../coda.app.mjs";
+
+export default {
+  key: "coda-create-rows",
+  name: "Create Rows",
+  description: `Inserts rows in a selected table.
+    More information at [Coda API](https://coda.io/developers/apis/v1#operation/upsertRows)`,
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/create-rows/create-rows.mjs:7]: Issue addressed

File: components/coda/actions/find-row/find-row.mjs [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+  description: "Searches for a row in the selected table using a column match search",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    coda,
+    docId: {
+      propDefinition: [
+        coda,
+        "docId",
+        (c) => c,
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/find-row/find-row.mjs:15]: Issue addressed

File: components/coda/actions/copy-doc/copy-doc.mjs [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+
+export default {
+  key: "coda-copy-doc",
+  name: "Copy Doc",
+  description: "Creates a copy of the specified doc",
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/copy-doc/copy-doc.mjs:6]: Addressed

File: components/coda/actions/find-row/find-row.mjs [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+      ],
+    },
+    query: {
+      propDefinition: [
+        coda,
+        "query",
+      ],
+      description: `Query used to filter returned rows, specified as \`<columnId>:"<value>"\`.
+        Example: \`query=c-tuVwxYz:"Apple"\`.
+        More information at [Coda API](https://coda.io/developers/apis/v1#operation/listRows)`,
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/find-row/find-row.mjs:44]: Addressed

[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/find-row/find-row.mjs:44]: Makes sense, it is better for the user. Addressed

File: components/coda/actions/find-row/find-row.mjs [hunk 28]
```diff
@@ -0,0 +1,127 @@
+import coda from "../../coda.app.mjs";
+
+export default {
+  key: "coda-find-row",
+  name: "Find Row",
+  description: "Searches for a row in the selected table using a column match search",
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/find-row/find-row.mjs:6]: Yep

File: components/coda/actions/list-docs/list-docs.mjs [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+    pageToken: {
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/list-docs/list-docs.mjs:69]: Addressed

File: components/coda/coda.app.mjs [hunk 30]
```diff
@@ -1,11 +1,303 @@
...
+  propDefinitions: {
+    title: {
+      type: "string",
+      label: "Title",
+      description: "Title of the doc. Defaults to `\"Untitled\"`",
+      optional: true,
+    },
+    docId: {
+      type: "string",
+      label: "Doc ID",
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/coda.app.mjs:15]: I've been thinking that users could also use variable references as inputs for ID props. In this case would it be better for them to know that the field is an ID?

File: components/coda/coda.app.mjs [hunk 31]
```diff
@@ -1,11 +1,303 @@
...
+      description: `ID of the doc
+        Options are displayed as \`{ "<docId>" : "<docValue>" }\``,
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/coda.app.mjs:17]: Addressed

File: components/coda/coda.app.mjs [hunk 32]
```diff
@@ -1,11 +1,303 @@
...
+    async listDocs(params = {}) {
+      let opts = {
+        path: "/docs",
+        params,
+      };
+      return await this._makeRequest(opts);
+    },
+    /**
+     * Lists tables in a doc according to parameters
+     * @param {object} [params]
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/coda.app.mjs:209]: Addressed

File: components/coda/coda.app.mjs [hunk 33]
```diff
@@ -1,11 +1,303 @@
...
+      opts.headers.Authorization = `Bearer ${this.$auth.api_token}`;
+      opts.headers["user-agent"] = "@PipedreamHQ/pipedream v0.1";
+      if (!opts.method) opts.method = "get";
+      const { path } = opts;
+      delete opts.path;
+      opts.url = `https://coda.io/apis/v1${path[0] === "/"
+        ? ""
+        : "/"}${path}`;
+      try {
+        return await axios(this, opts);
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/coda.app.mjs:152]: Addressed

File: components/coda/actions/find-row/find-row.mjs [hunk 34]
```diff
@@ -0,0 +1,127 @@
...
+    tableId: {
+      propDefinition: [
+        coda,
+        "tableId",
+        (c) => ({
+          docId: c.docId,
+        }),
+      ],
+    },
+    columnId: {
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/find-row/find-row.mjs:27]: Addressed

File: components/coda/coda.app.mjs [hunk 35]
```diff
@@ -1,11 +1,303 @@
...
+    docId: {
+      type: "string",
+      label: "Doc ID",
+      description: `ID of the doc
+        Options are displayed as \`{ "<docId>" : "<docValue>" }\``,
+      async options () {
+        return this._makeOptionsResponse(
+          (await this.listDocs()).items,
+        );
+      },
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/coda.app.mjs:22]: Addressed

File: components/coda/actions/update-row/update-row.mjs [hunk 36]
```diff
@@ -0,0 +1,80 @@
...
+      propDefinition: [
+        coda,
+        "disableParsing",
+      ],
+    },
+    row: {
+      type: "string",
+      label: "Row",
+      description: "An edit made to a particular row",
+    },
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/update-row/update-row.mjs:58]: Thanks for the suggestion!

File: components/coda/actions/list-docs/list-docs.mjs (hunk 22 above)
Comment (andrewjschuang) [reply components/coda/actions/list-docs/list-docs.mjs:68]: Addressed

File: components/coda/actions/find-row/find-row.mjs [hunk 37]
```diff
@@ -0,0 +1,127 @@
...
+        "rich",
+      ],
+    },
+    limit: {
+      propDefinition: [
+        coda,
+        "limit",
+      ],
+    },
+    pageToken: {
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/find-row/find-row.mjs]: Addressed

File: components/coda/actions/list-tables/list-tables.mjs [hunk 38]
```diff
@@ -0,0 +1,63 @@
...
+  },
+  async run({ $ }) {
+    let params = {
+      sortBy: this.sortBy,
+      tableTypes: this.tableTypes.toString(),
+      limit: this.limit,
+      pageToken: this.pageToken,
+    };
+
+    let response = await this.coda.listTables(
```
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/coda/actions/list-tables/list-tables.mjs]: Addressed

File: components/coda/coda.app.mjs [hunk 39]
```diff
@@ -1,11 +1,307 @@
...
+          context,
+        } = this._makeOptionsResponse(response);
+
+        return {
+          options: options
+            .map((row) => ({
+              label: `${counter++}: ${row.label}`,
+              value: row.value,
+            })),
+          context,
```
Comment (js07) [inline components/coda/coda.app.mjs:68]: Since the `counter` variable is set to `0` each time this `options` functions is called, each page of rows will have row numbers starting at `0`.

To maintain the counter across `options` invocations, you can pass `counter` through [`context`](https://pipedream.com/docs/components/api/#component-api). For example:

```suggestion
        let { counter = 0 } = prevContext;
        const response = await this.findRow(this, docId, tableId, {
          pageToken: prevContext.nextPageToken,
          sortBy: "natural",
        });

        const {
          options,
          context,
        } = this._makeOptionsResponse(response);

        return {
          options: options
            .map((row) => ({
              label: `${counter++}: ${row.label}`,
              value: row.value,
            })),
          context: {
            ...context,
            counter,
          },
```

File: components/coda/actions/list-columns/list-columns.mjs [hunk 40]
```diff
@@ -30,33 +29,37 @@ export default {
...
 
-    $.export("$summary", `Retrieved ${response.items.length} column(s)`);
-    return response;
+    if (items.length > this.max) items.length = this.max;
+
+    $.export("$summary", `Retrieved ${items.length} column(s)`);
+
+    return {
+      items,
+    };
```
Comment (js07) [inline components/coda/actions/list-columns/list-columns.mjs:63]: The return value from a "list" action is typically an array of the items being listed, to simplify using the results from actions in workflows. Some examples in existing components: [Airtable](https://github.com/PipedreamHQ/pipedream/blob/78b07a8bc24adb1724467987e85743fa7bed711c/components/airtable/actions/common-list.js#L54), [Google Sheets](https://github.com/PipedreamHQ/pipedream/blob/1dcc47febdc68b2b9d1990c3d7658f91a0d1f090/components/google_sheets/actions/list-worksheets/list-worksheets.mjs#L30), [Eventbrite](https://github.com/PipedreamHQ/pipedream/blob/78b07a8bc24adb1724467987e85743fa7bed711c/components/eventbrite/actions/get-event-attendees/get-event-attendees.js#L41)

```suggestion
    return items;
```

Also applies to `list-docs.mjs` and `list-tables.mjs`.

**Review by js07 — COMMENTED** [review]
The changes look good! I just made a couple minor comments.

File: components/coda/actions/list-columns/list-columns.mjs (hunk 40 above)
Comment (andrewjschuang) [reply components/coda/actions/list-columns/list-columns.mjs:63]: Addressed

File: components/coda/coda.app.mjs (hunk 39 above)
Comment (andrewjschuang) [reply components/coda/coda.app.mjs:68]: Addressed

**Comment by andrewjschuang** [comment]
Thanks, @js07! I've applied the requested changes.
I think I prefer rebasing on top of [`PipdreamHQ:master`](https://github.com/PipedreamHQ/pipedream); although it creates new commits, I think it looks a bit cleaner in the tree graph

**Comment by dannyroosevelt** [comment]
@andrewjschuang go ahead and merge!
